File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.js
const Show = {
    Enter: "님이 들어왔습니다.",
    Leave: "님이 나갔습니다."
};

function solution(record) {
    const nicknName = {};
    const output = [];

    record.forEach((e) => {
        const [execution, id, name] = e.split(' ');

        if (["Enter", "Change"].includes(execution)) nicknName[id] = name;
        if (["Enter", "Leave"].includes(execution)) output.push([execution, id]);
    });
    return output.map(([execution, id]) => `${nicknName[id]}${Show[execution]}`);
}

File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <sstream>
#include <list>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    unordered_map<string, string> m;
    list<pair<string, string>> li;

    stringstream ss;
    for (const auto& data : record) {
        ss.str(data);
        string command, userID, nickName; ss >> command >> userID >> nickName;
        ss.clear();

        if (command != "Change") li.emplace_back(make_pair(command, userID));
        if (command != "Leave") m[userID] = nickName;
    }

    for (const auto& pair : li) {
        const auto& command = pair.first, userID = pair.second;

        if (command == "Enter") answer.emplace_back(m[userID] + "님이 들어왔습니다.");
        else answer.emplace_back(m[userID] + "님이 나갔습니다.");
    }

    return answer;
}